### PR TITLE
PWGHF: D+h correlations  -  efficiency correction

### PR DESCRIPTION
**Pull request by apalasciano**
D+ - h correlation task:
- efficiency evaluation
- efficiency correction to correl distribution
- QA histogram
- removing obsolete code

### CODE REVIEW

File: PWGHF/HFC/Tasks/taskCorrelationDplusHadrons.cxx [hunk 1]
```diff
@@ -78,65 +69,126 @@ std::vector<double> efficiencyDmeson(npTBinsEfficiency + 1);
...
+    AxisSpec axisPtD = {(std::vector<double>)binsPtEfficiencyD, "#it{p}_{T}^{D} (GeV/#it{c})"};
+    AxisSpec axisMultFT0M = {binsMultFT0M, "MultiplicityFT0M"};
+    AxisSpec axisDeltaEta = {binsEta, "#it{#eta}^{Hadron}-#it{#eta}^{D}"};
+    AxisSpec axisDeltaPhi = {binsPhi, "#it{#varphi}^{Hadron}-#it{#varphi}^{D} (rad)"};
+    AxisSpec axisPtHadron = {(std::vector<double>)binsPtHadron, "#it{p}_{T}^{Hadron} (GeV/#it{c})"};
+    AxisSpec axisPoolBin = {binsPoolBin, "poolBin"};
+    AxisSpec axisDplusPrompt = {2, -0.5, 1.5, "Prompt D+"};
+
+    registry.add("hDeltaEtaPtIntSignalRegion", stringDHadron + stringSignal + stringDeltaEta + "entries", {HistType::kTH1F, {{axisDeltaEta}}}),
+    registry.add("hDeltaPhiPtIntSignalRegion", stringDHadron + stringSignal + stringDeltaPhi + "entries", {HistType::kTH1F, {{axisDeltaPhi}}}),
```
Comment (vkucera) [inline PWGHF/HFC/Tasks/taskCorrelationDplusHadrons.cxx:132]: This won't compile. Please make sure your branch compiles before making a PR.
```suggestion
    registry.add("hDeltaEtaPtIntSignalRegion", stringDHadron + stringSignal + stringDeltaEta + "entries", {HistType::kTH1F, {{axisDeltaEta}}});
    registry.add("hDeltaPhiPtIntSignalRegion", stringDHadron + stringSignal + stringDeltaPhi + "entries", {HistType::kTH1F, {{axisDeltaPhi}}});
```

**Review by vkucera — CHANGES_REQUESTED** [review]

File: PWGHF/HFC/Tasks/taskCorrelationDplusHadrons.cxx (hunk 1 above)
Comment (vkucera) [reply PWGHF/HFC/Tasks/taskCorrelationDplusHadrons.cxx]: Sorry, it does compile. It just does something slightly different. :-)

File: PWGHF/HFC/Tasks/taskCorrelationDplusHadrons.cxx [hunk 2]
```diff
@@ -78,65 +69,126 @@ std::vector<double> efficiencyDmeson(npTBinsEfficiency + 1);
...
+                       kCandidateStepMcReco,
+                       kCandidateStepMcRecoInAcceptance,
+                       kCandidateNSteps };
+
+  using CandDplusMcReco = soa::Filtered<soa::Join<aod::HfCand3Prong, aod::HfSelDplusToPiKPi, aod::HfCand3ProngMcRec>>;
+  using CandDplusMcGen = soa::Join<aod::McParticles, aod::HfCand3ProngMcGen>;
+
+  HistogramRegistry registry{"registry", {}, OutputObjHandlingPolicy::AnalysisObject};
+
+  Filter dplusFilter = ((o2::aod::hf_track_index::hfflag & static_cast<uint8_t>(1 << aod::hf_cand_3prong::DecayType::DplusToPiKPi)) != static_cast<uint8_t>(0)) && aod::hf_sel_candidate_dplus::isSelDplusToPiKPi >= selectionFlagDplus;
```
Comment (vkucera) [inline PWGHF/HFC/Tasks/taskCorrelationDplusHadrons.cxx:117]: Please keep the registry last.

File: PWGHF/HFC/TableProducer/correlatorDplusHadrons.cxx [hunk 3]
```diff
@@ -141,6 +141,8 @@ struct HfCorrelatorDplusHadrons {
 
   Configurable<int> selectionFlagDplus{"selectionFlagDplus", 7, "Selection Flag for Dplus"}; // 7 corresponds to topo+PID cuts
   Configurable<int> applyEfficiency{"applyEfficiency", 1, "Flag for applying D-meson efficiency weights"};
+  Configurable<std::vector<double>> binsPtEfficiencyD{"binsPtEfficiencyD", std::vector<double>{o2::analysis::hf_cuts_dplus_to_pi_k_pi::vecBinsPt}, "pT bin limits for efficiency"};
+  Configurable<std::vector<double>> efficiencyD{"efficiencyD", {1., 1., 1., 1., 1., 1.}, "efficiency values for D+ meson"};
```
Comment (vkucera) [inline PWGHF/HFC/TableProducer/correlatorDplusHadrons.cxx:145]:
```suggestion
  Configurable<std::vector<float>> efficiencyD{"efficiencyD", {1., 1., 1., 1., 1., 1.}, "efficiency values for D+ meson"};
```

File: PWGHF/HFC/Tasks/taskCorrelationDplusHadrons.cxx [hunk 4]
```diff
@@ -78,65 +69,126 @@ std::vector<double> efficiencyDmeson(npTBinsEfficiency + 1);
...
   Configurable<std::vector<double>> binsPtCorrelations{"binsPtCorrelations", std::vector<double>{pTBinsCorrelations_v}, "pT bin limits for correlation plots"};
-  Configurable<std::vector<double>> binsPtEfficiency{"binsPtEfficiency", std::vector<double>{o2::analysis::hf_cuts_dplus_to_pi_k_pi::vecBinsPt}, "pT bin limits for efficiency"};
+  Configurable<std::vector<double>> binsPtHadron{"binsPtHadron", std::vector<double>{0.3, 2., 4., 8., 12., 50.}, "pT bin limits for assoc particle efficiency"};
+  Configurable<std::vector<double>> binsPtEfficiencyD{"binsPtEfficiencyD", std::vector<double>{o2::analysis::hf_cuts_dplus_to_pi_k_pi::vecBinsPt}, "pT bin limits for efficiency"};
+  Configurable<std::vector<double>> binsPtEfficiencyHad{"binsPtEfficiencyHad", std::vector<double>{0.3, 2., 4., 8., 12., 50.}, "pT bin limits for associated particle efficiency"};
+  Configurable<std::vector<double>> efficiencyD{"efficiencyD", {1., 1., 1., 1., 1., 1.}, "efficiency values for D+ meson"};
+  Configurable<std::vector<double>> efficiencyHad{"efficiencyHad", {1., 1., 1., 1., 1., 1.}, "efficiency values for associated particles"};
```
Comment (vkucera) [inline PWGHF/HFC/Tasks/taskCorrelationDplusHadrons.cxx:79]:
```suggestion
  Configurable<std::vector<float>> efficiencyD{"efficiencyD", {1., 1., 1., 1., 1., 1.}, "efficiency values for D+ meson"};
  Configurable<std::vector<float>> efficiencyHad{"efficiencyHad", {1., 1., 1., 1., 1., 1.}, "efficiency values for associated particles"};
```

**Comment by fcolamar** [comment]
Hi Antonio, I'm generally ok with this PR. As discussed in private we can refine the associated pt ranges,.but this can be done at a later stage.
If there are no further comments by @vkucera or the others, I would approve this PR, since we would like to have a train run on the updated code and get some results to compare with Samuele's Ds-h before Monday's HF approvals of the latter study.
Cheers,
Fabio

**Review by fcolamar — APPROVED** [review]

**Pull request merged**